File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/IndexBiMap.java
package net.namekdev.entity_tracker.utils;

public class IndexBiMap extends Array<Integer> {
	public IndexBiMap() {
		super();
	}

	public IndexBiMap(int capacity) {
		super(capacity);
	}

	public int getLocalIndex(int globalIndex) {
		for (int i = 0, n = _array.size(); i < n; ++i) {
			Integer val = _array.get(i);

			if (val == null) {
				break;
			}

			if (val == globalIndex) {
				return i;
			}
		}

		return -1;
	}

	public int getGlobalIndex(int localIndex) {
		return _array.get(localIndex);
	}
}

File: src/commonMain/kotlin/net/namekdev/entity_tracker/utils/IndexBiMap.kt
package net.namekdev.entity_tracker.utils

class IndexBiMap : AutoSizedArray<Int> {
    constructor() : super() {}

    constructor(capacity: Int) : super(capacity) {}

    fun getLocalIndex(globalIndex: Int): Int {
        var i = 0
        val n = _array.size
        while (i < n) {
            val value = _array[i] ?: break

            if (value === globalIndex) {
                return i
            }
            ++i
        }

        return -1
    }

    fun getGlobalIndex(localIndex: Int): Int {
        return _array[localIndex]!!
    }
}
